edward/inferences/klqp.py

```diff
@@ -1127,3 +1184,90 @@ def build_score_rb_loss_and_gradients(inference, var_list):
+    """
+    """
+    rej_samplers = {
+      Gamma: GammaRejectionSampler
+    }
+
+    rep = [0.0] * inference.n_samples
+    cor = [0.0] * inference.n_samples
+    base_scope = tf.get_default_graph().unique_name("inference") + '/'
+    for s in range(inference.n_samples):
+      # Form dictionary in order to replace conditioning on prior or
+      # observed variable with conditioning on a specific value.
+      scope = base_scope + tf.get_default_graph().unique_name("sample")
+      dict_swap = {}
+      for x, qx in six.iteritems(inference.data):
+        if isinstance(x, RandomVariable):
+          if isinstance(qx, RandomVariable):
+            qx_copy = copy(qx, scope=scope)
+            dict_swap[x] = qx_copy.value()
+          else:
+            dict_swap[x] = qx
+
+      p_log_prob = 0.
+      q_log_prob = 0.
+      r_log_prob = 0.
+
+      for z, qz in six.iteritems(inference.latent_vars):
+        # Copy q(z) to obtain new set of posterior samples.
+        qz_copy = copy(qz, scope=scope)
+        sampler = rej_samplers[qz_copy.__class__](density=qz)
+
+        if epsilon is not None:  # temporary
+          pass
+        else:
+          dict_swap[z] = qz_copy.value()
+          print('sample:', dict_swap[z])
+          epsilon = sampler.h_inverse(dict_swap[z])
+
+        dict_swap[z] = sampler.h(epsilon)
+        q_log_prob += tf.reduce_sum(
+          inference.scale.get(z, 1.0) * qz_copy.log_prob(dict_swap[z]))
+        r_log_prob += -tf.log(tf.gradients(dict_swap[z], epsilon))
+
+      for z in six.iterkeys(inference.latent_vars):
+        z_copy = copy(z, dict_swap, scope=scope)
+        p_log_prob += tf.reduce_sum(
+          inference.scale.get(z, 1.0) * z_copy.log_prob(dict_swap[z]))
+
+      for x in six.iterkeys(inference.data):
+        if isinstance(x, RandomVariable):
+          x_copy = copy(x, dict_swap, scope=scope)
+          p_log_prob += tf.reduce_sum(
+            inference.scale.get(x, 1.0) * x_copy.log_prob(dict_swap[x]))
+
+      rep[s] = p_log_prob
+      cor[s] = tf.stop_gradient(p_log_prob) * (q_log_prob - r_log_prob)
+
+    rep = tf.reduce_mean(rep)
+    cor = tf.reduce_mean(cor)
+    q_entropy = tf.reduce_sum([
+      tf.reduce_sum(qz.entropy())
+      for z, qz in six.iteritems(inference.latent_vars)])
+    reg_penalty = tf.reduce_sum(tf.losses.get_regularization_losses())
+
+    loss = -(rep + q_entropy - reg_penalty)
+
+    if inference.logging:
+      tf.summary.scalar("loss/reparam_objective", rep,
+                        collections=[inference._summary_key])
+      tf.summary.scalar("loss/correction_term", cor,
+                        collections=[inference._summary_key])
+      tf.summary.scalar("loss/q_entropy", q_entropy,
+                        collections=[inference._summary_key])
+      tf.summary.scalar("loss/reg_penalty", reg_penalty,
+                        collections=[inference._summary_key])
+
+    g_rep = tf.gradients(rep, var_list)
```

Comment:
Can you explain why you need the multiple gradient calls and not just one? This seems inefficient.